lib/stack_master/template_compilers/sparkle_formation.rb

```diff
@@ -1,18 +1,52 @@
+require_relative '../sparkle_formation/compile_time/parameters_validator'
+require_relative '../sparkle_formation/compile_time/definitions_validator'
+require_relative '../sparkle_formation/compile_time/state_builder'
+
 module StackMaster::TemplateCompilers
   class SparkleFormation
+
+    CompileTime = StackMaster::SparkleFormation::CompileTime
+
     def self.require_dependencies
       require 'sparkle_formation'
       require 'stack_master/sparkle_formation/template_file'
     end
 
-    def self.compile(template_file_path, compiler_options = {})
-      if compiler_options["sparkle_path"]
-        ::SparkleFormation.sparkle_path = File.expand_path(compiler_options["sparkle_path"])
+    def self.compile(template_file_path, parameters, compiler_options = {})
+      if compiler_options['sparkle_path']
+        ::SparkleFormation.sparkle_path = File.expand_path(compiler_options['sparkle_path'])
       else
         ::SparkleFormation.sparkle_path = File.dirname(template_file_path)
       end
+      sparkle_template = ::SparkleFormation.compile(template_file_path, :sparkle)
+      definitions = sparkle_template.parameters
+      validate_definitions(definitions)
+      validate_parameters(definitions, parameters)
+
+      sparkle_template.compile_time_parameter_setter do
+        sparkle_template.compile_state = create_state(definitions, parameters)
+        remove_compile_parameters(definitions, parameters)
+      end
+
+      JSON.pretty_generate(sparkle_template)
+    end
+
+    private
+
+    def self.validate_definitions(definitions)
+      CompileTime::DefinitionsValidator.new(definitions).validate
+    end
+
+    def self.validate_parameters(definitions, parameters)
+      CompileTime::ParametersValidator.new(definitions, parameters).validate
+    end
+
+    def self.create_state(definitions, parameters)
+      CompileTime::StateBuilder.new(definitions, parameters).build
+    end
 
-      JSON.pretty_generate(::SparkleFormation.compile(template_file_path))
+    def self.remove_compile_parameters(definitions, parameters)
+      definitions.each {|name, _definition| parameters.delete(name.to_s.camelize)}
```

Comment:
Is the assumption that compile time parameters will never be used in the template? I'm not sure that's true....

Reply:
Seems odd to use them in two ways...to us. Why you use it as a compile time parameter and also as a normal parameter. 
If you only use it as a compile time parameter its going to force you include its as a normal parameter whether you use it or not

Reply:
A parameter can be used in multiple ways, to generate some resources and pass into user data. I don't have a solid use case though so lets see how it goes

Reply:
Both to and against can be argued for. I reckon in a lot of cases compile time parameters will be used to programmatically construct CFn templates, and having to include them in the resulting template will more often than not pollute the paramters file, and make it less readable. We have just been opinionated here.